pkg/security/admission/admission.go

```diff
@@ -0,0 +1,385 @@
+		// we will grab the old SC of the container, insert our generated SC and validate.
+		// We will then restore the old SCC.  If we validate the entire pod then we will set
+		// the container SCs back to the generated SC at the end of this method.
+		originalSC := c.SecurityContext
+		c.SecurityContext = sc
+		providerErrs := provider.ValidateSecurityContext(pod, &c)
+		c.SecurityContext = originalSC
+
+		// now that we've restored the old sc check for errors, if everything is ok we'll apply
+		// the generated sc later
+		if len(providerErrs) > 0 {
+			return providerErrs
+		}
+	}
+
+	// if we've reached this code then we've generated and validated an SC for every container in the
+	// pod so let's apply what we generated
+	for i, sc := range generatedSCs {
+		pod.Spec.Containers[i].SecurityContext = sc
+	}
+	return nil
+}
+
+// createProvidersFromConstraints creates providers from the constraints supplied, including
+// looking up pre-allocated values if necessary using the pod's namespace.
+func (c *constraint) createProvidersFromConstraints(ns string, sccs []*kapi.SecurityContextConstraints) ([]scc.SecurityContextConstraintsProvider, error) {
+	providers := make([]scc.SecurityContextConstraintsProvider, 0)
+	// namespace is declared here for reuse but we will not fetch it unless required by
+	// the matched constraints
+	var namespace *kapi.Namespace
+	var err error
+
+	// set pre-allocated values on constraints.  By placing them in a map we also de-dupe the constraints
+	// that match the sa and user.
+	for _, constraint := range sccs {
+		if requiresPreAllocatedUIDRange(constraint) {
+			namespace, err = c.getNamespace(ns, namespace)
+			if err != nil {
+				return nil, err
+			}
+
+			min, max, err := getPreallocatedUIDRange(namespace)
+			if err != nil {
+				return nil, fmt.Errorf("unable to find pre-allocated uid annotation for namespace %s while trying to configure SCC %s: %v", namespace.Name, constraint.Name, err)
+			}
+			constraint.RunAsUser.UIDRangeMin = min
+			constraint.RunAsUser.UIDRangeMax = max
+		}
+
+		if requiresPreAllocatedSELinuxLevel(constraint) {
+			namespace, err = c.getNamespace(ns, namespace)
+			if err != nil {
+				return nil, err
+			}
+
+			// if we don't have options create them here so we can add the preallocated settings
+			if constraint.SELinuxContext.SELinuxOptions == nil {
+				constraint.SELinuxContext.SELinuxOptions = &kapi.SELinuxOptions{}
+			}
+
+			level, err := getPreallocatedLevel(namespace)
+			if err != nil {
+				return nil, fmt.Errorf("unable to find pre-allocated mcs annotation for namespace %s while trying to configure SCC %s: %v", namespace.Name, constraint.Name, err)
+			}
+			constraint.SELinuxContext.SELinuxOptions.Level = level
+		}
+
+		provider, err := scc.NewSimpleProvider(constraint)
+		if err != nil {
+			return nil, fmt.Errorf("error creating provider for SCC %s in namespace %s: %v", constraint.Name, ns, err)
```

Comment:
Do you need to use a special type of error here, since the admission controllers deal in validation-like errors?